opentelemetry-sdk/src/logs/record.rs

```diff
@@ -0,0 +1,203 @@
+use opentelemetry::{
+    logs::{AnyValue, Severity},
+    trace::{SpanContext, SpanId, TraceFlags, TraceId},
+    Key,
+};
+use std::{borrow::Cow, time::SystemTime};
+
+#[derive(Debug, Clone, Default)]
+#[non_exhaustive]
+/// LogRecord represents all data carried by a log record, and
+/// is provided to `LogExporter`s as input.
+pub struct LogRecord {
+    /// Record timestamp
+    pub timestamp: Option<SystemTime>,
+
+    /// Timestamp for when the record was observed by OpenTelemetry
+    pub observed_timestamp: Option<SystemTime>,
+
+    /// Trace context for logs associated with spans
+    pub trace_context: Option<TraceContext>,
+
+    /// The original severity string from the source
+    pub severity_text: Option<Cow<'static, str>>,
+    /// The corresponding severity value, normalized
+    pub severity_number: Option<Severity>,
+
+    /// Record body
+    pub body: Option<AnyValue>,
+
+    /// Additional attributes associated with this record
+    pub attributes: Option<Vec<(Key, AnyValue)>>,
```

Comment:
nit: is there any recommended way of using Option<None> vs empty vec! here?

Reply:
I felt Option would be better, specifically for the no attributes scenario. The empty vector still has some minimal heap presence, along with memory required for storing length and capacity.

Reply:
From the [docs](https://doc.rust-lang.org/stable/std/vec/struct.Vec.html#method.new), it looks like empty vector does not allocate:

> The vector will not allocate until elements are pushed onto it.

Reply:
Yes, I read somewhere that some minimal bookkeeping allocation done on heap initially, but this may be wrong. Irrespective of that, I think enum allocation for None would be minimal as compared to the empty vector :)

Reply:
The allocation for the `Option` should be same regardless of whether it was `None` or `Some`, right? It should just depend on the type `T` that's provided to the `Option`.

```rust
let option_none: Option<Vec<(&str, u32)>> = None;

let mut vec = Vec::<(&str, u32)>::new();
vec.push(("a", 1));
vec.push(("b", 2));
vec.push(("c", 3));

let option_some = Some(vec);

println!("{}", mem::size_of::<Option<Vec<(&str, u32)>>>()); //24
println!("{}", mem::size_of::<Vec::<(&str, u32)>>()); //24
println!("{}", mem::size_of_val(&option_none)); //24
println!("{}", mem::size_of_val(&option_some)); //24
```

Reply:
> But happy to change if you feel otherwise :)

I'm not sure if using `Option` is the more idiomatic thing but I think `Option` adds another layer of indirection to check if there were any attributes present. Even if the `Option` is a `Some` variant, it could have an empty vector wrapped in `Some`. With using `Vec` as the type for the `attributes` field, we don't have this indirection. Other than that, I don't know of any other benefit to using `Vec` here 😄

Reply:
Good points. I guess one of the benefits of moving LogRecord to SDK is that we can change this internal representations afterwards too :)